File: src/AdventOfCode.Site/scripts/ts/puzzles/2022/Day08.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { Point } from '../Point';
import { Size } from '../Size';
import { Puzzle2022 } from './Puzzle2022';

export class Day08 extends Puzzle2022 {
    visibleTrees: number;
    maximumScenicScore: number;

    override get name() {
        return 'Treetop Tree House';
    }

    override get day() {
        return 8;
    }

    protected override get requiresData() {
        return true;
    }

    static countVisibleTrees(grid: string[]): [visibleTrees: number, maximumScenicScore: number] {
        const up = new Size(0, -1);
        const down = new Size(0, 1);
        const left = new Size(-1, 0);
        const right = new Size(1, 0);

        const parse = (grid: string[]): [trees: Map<string, Tree>, height: number, width: number] => {
            const height = grid.length;
            const width = grid[0].length;

            const trees = new Map<string, Tree>();

            for (let y = 0; y < height; y++) {
                for (let x = 0; x < width; x++) {
                    const location = new Point(x, y);
                    trees.set(location.asString(), new Tree(location, grid[y][x].charCodeAt(0) - '0'.charCodeAt(0)));
                }
            }

            return [trees, height, width];
        };

        const sweep = (origin: Point, direction: Size, trees: Map<string, Tree>) => {
            let location = origin;
            let current = trees.get(location.asString());
            current.markVisible();

            while (trees.has((location = Point.add(location, direction)).asString())) {
                const next = trees.get(location.asString());
                if (next.height > current.height) {
                    next.markVisible();
                    current = next;
                }
            }
        };

        const countVisibleTrees = (trees: Map<string, Tree>, height: number, width: number) => {
            const topLeft = new Point(0, 0);
            const topRight = new Point(width - 1, 0);
            const bottomLeft = new Point(width - 1, height - 1);

            const directions: [origin: Point, next: Size, forward: Size][] = [
                [topLeft, down, right], // Left to right
                [topRight, down, left], // Right to left
                [topLeft, right, down], // Top to bottom
                [bottomLeft, left, up], // Bottom to top
            ];

            for (const [first, next, direction] of directions) {
                let origin = first;
                do {
                    sweep(origin, direction, trees);
                    origin = Point.add(origin, next);
                } while (trees.has(origin.asString()));
            }

            let count = 0;

            for (const tree of trees.values()) {
                if (tree.isVisible) {
                    count++;
                }
            }

            return count;
        };

        const score = (origin: Tree, direction: Size, trees: Map<string, Tree>) => {
            let location = origin.location;
            let score = 0;

            while (trees.has((location = Point.add(location, direction)).asString())) {
                const next = trees.get(location.asString());
                score++;

                if (next.height >= origin.height) {
                    break;
                }
            }

            return score;
        };

        const getMaximumScenicScore = (trees: Map<string, Tree>, height: number, width: number) => {
            const directions = [up, down, left, right];

            let maximum = 0;

            for (let y = 0; y < height; y++) {
                for (let x = 0; x < width; x++) {
                    const tree = trees.get(new Point(x, y).asString());

                    if (tree.isVisible) {
                        let scenicScore = 1;

                        for (const direction of directions) {
                            scenicScore *= score(tree, direction, trees);
                        }

                        maximum = Math.max(scenicScore, maximum);
                    }
                }
            }

            return maximum;
        };

        const [trees, height, width] = parse(grid);
        const count = countVisibleTrees(trees, height, width);
        const maximum = getMaximumScenicScore(trees, height, width);

        return [count, maximum];
    }

    override solveCore(_: string[]) {
        const grid = this.readResourceAsLines();

        [this.visibleTrees, this.maximumScenicScore] = Day08.countVisibleTrees(grid);

        console.info(`There are ${this.visibleTrees} trees visible from outside the grid.`);
        console.info(`The highest scenic score is ${this.maximumScenicScore}.`);

        return this.createResult([this.visibleTrees, this.maximumScenicScore]);
    }
}

class Tree {
    private isTreeVisible: boolean;

    constructor(public readonly location: Point, public readonly height: number) {
        this.isTreeVisible = false;
    }

    public get isVisible() {
        return this.isTreeVisible;
    }

    public markVisible() {
        this.isTreeVisible = true;
    }
}

File: src/AdventOfCode/Puzzles/Y2022/Day08.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2022;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2022/day/8</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2022, 08, "Treetop Tree House", RequiresData = true)]
public sealed class Day08 : Puzzle
{
    /// <summary>
    /// Gets how many trees are visible from outside the grid.
    /// </summary>
    public int VisibleTrees { get; private set; }

    /// <summary>
    /// Gets the maximum scenic score that can be achieved for a treehouse in a tree in the grid.
    /// </summary>
    public int MaximumScenicScore { get; private set; }

    /// <summary>
    /// Returns how many trees are visible from outside the specified grid.
    /// </summary>
    /// <param name="grid">The grid of trees and heights.</param>
    /// <returns>
    /// How many trees are visible from outside the grid specified by <paramref name="grid"/>
    /// and the maximum scenic score that can be achieved for a treehouse in a tree in the grid.
    /// </returns>
    public static (int VisibleTrees, int MaximumScenicScore) CountVisibleTrees(IList<string> grid)
    {
        var trees = Parse(grid);
        int count = CountVisibleTrees(trees);
        int score = GetMaximumScenicScore(trees);

        return (count, score);

        static Dictionary<Point, Tree> Parse(IList<string> grid)
        {
            int height = grid.Count;
            int width = grid[0].Length;

            var trees = new Dictionary<Point, Tree>(height * width);

            for (int y = 0; y < height; y++)
            {
                for (int x = 0; x < width; x++)
                {
                    var location = new Point(x, y);
                    trees[location] = new Tree(location, grid[y][x] - '0');
                }
            }

            return trees;
        }

        static int CountVisibleTrees(Dictionary<Point, Tree> trees)
        {
            int height = trees.Keys.Max((p) => p.Y) + 1;
            int width = trees.Keys.Max((p) => p.X) + 1;

            var topLeft = new Point(0, 0);
            var topRight = new Point(width - 1, 0);
            var bottomLeft = new Point(width - 1, height - 1);

            var directions = new (Point Origin, Size Next, Size Forward)[]
            {
                (topLeft,    Directions.Down,  Directions.Right), // Left to right
                (topRight,   Directions.Down,  Directions.Left), // Right to left
                (topLeft,    Directions.Right, Directions.Down), // Top to bottom
                (bottomLeft, Directions.Left,  Directions.Up), // Bottom to top
            };

            foreach (var (first, next, direction) in directions)
            {
                var origin = first;

                do
                {
                    Sweep(origin, direction, trees);
                    origin += next;
                }
                while (trees.ContainsKey(origin));
            }

            return trees.Count((p) => p.Value.IsVisible);
        }

        static void Sweep(Point origin, Size direction, Dictionary<Point, Tree> trees)
        {
            var location = origin;
            var current = trees[location];
            current.MarkVisible();

            while (trees.TryGetValue(location += direction, out var next))
            {
                if (next.Height > current.Height)
                {
                    next.MarkVisible();
                    current = next;
                }
            }
        }

        static int Score(Tree origin, Size direction, Dictionary<Point, Tree> trees)
        {
            var location = origin.Location;

            int score = 0;

            while (trees.TryGetValue(location += direction, out var next))
            {
                score++;

                if (next.Height >= origin.Height)
                {
                    break;
                }
            }

            return score;
        }

        static int GetMaximumScenicScore(Dictionary<Point, Tree> trees)
        {
            int height = trees.Keys.Max((p) => p.Y) + 1;
            int width = trees.Keys.Max((p) => p.X) + 1;

            var directions = new[]
            {
                Directions.Up,
                Directions.Down,
                Directions.Left,
                Directions.Right,
            };

            int maximum = 0;

            for (int y = 0; y < height; y++)
            {
                for (int x = 0; x < width; x++)
                {
                    var tree = trees[new(x, y)];

                    if (tree.IsVisible)
                    {
                        int scenicScore = 1;

                        foreach (var direction in directions)
                        {
                            scenicScore *= Score(tree, direction, trees);
                        }

                        maximum = Math.Max(scenicScore, maximum);
                    }
                }
            }

            return maximum;
        }
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        var grid = await ReadResourceAsLinesAsync(cancellationToken);

        (VisibleTrees, MaximumScenicScore) = CountVisibleTrees(grid);

        if (Verbose)
        {
            Logger.WriteLine("There are {0} trees visible from outside the grid.", VisibleTrees);
            Logger.WriteLine("The highest scenic score is {0}.", MaximumScenicScore);
        }

        return PuzzleResult.Create(VisibleTrees, MaximumScenicScore);
    }

    private sealed record Tree(Point Location, int Height)
    {
        public bool IsVisible { get; private set; }

        public void MarkVisible() => IsVisible = true;
    }
}
